rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -3278,8 +3278,9 @@ static void message_info_from_sample_info(
     message_info->publisher_gid.data, &info.publication_handle,
     sizeof(info.publication_handle));
   message_info->source_timestamp = info.source_timestamp;
-  // TODO(iluetkeb) add received timestamp, when implemented by Cyclone
-  message_info->received_timestamp = 0;
+  // TODO(iluetkeb) get received_timestamp from Cyclone when implemented there
+  message_info->received_timestamp = std::chrono::duration_cast<std::chrono::nanoseconds>(
+    std::chrono::steady_clock::now().time_since_epoch()).count();
```

Comment:
I think this should be the system clock. Otherwise it is not relatable to other time stamps from other runs. And it also makes it impossible to properly compare it to a timestamp on the sending side (in order to calculate transit time of the message over the network).

Reply:
@wjwwood You opened up a box of "pandora". :face_with_head_bandage: 
Actually, I was under the impression that construction like this. 
```
std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now().time_since_epoch()).count()
```
Will give us the number of nanoseconds since the Unix epoch, i.e., Jan 1 of, 1970. 
However, it seems this is not true, and it will give us a number of nanoseconds since the `steady_clock` start - which is the **number of nanoseconds since the boot time**. And you are right in this case - this timestamp is out of context in regards to the world clock time.
Initially, I chose `steady_clock` because it is not adjusting and monotonic.

On the other side, using `system_clock` also has its own drawbacks. I found a good blogpost with explanations and test examples: https://www.modernescpp.com/index.php/the-three-clocks/
1. The `system_clock` can jump backward after adjustment or clock synch. This could lead to the wrong order during replay.
2. The `system_clock` on Windows specifically has microseconds in resolution. This is likely why `rmw_connextdds` doesn't support sent and received timestmaps on Windows https://github.com/ros2/rmw_connextdds/blob/d2bd90826e19097a429ba150b12e67d3822c1ccc/rmw_connextdds_common/src/common/rmw_impl.cpp#L1970-L1983
  ```cpp
  rmw_connextdds_message_info_from_dds(
    rmw_message_info_t * const to,
    const DDS_SampleInfo * const from)
  {
    rmw_connextdds_ih_to_gid(from->publication_handle, to->publisher_gid);
  // Message timestamps are disabled on Windows because RTI Connext DDS
  // does not support a high enough clock resolution by default (see: _ftime()).
  #if !RTI_WIN32
    to->source_timestamp = dds_time_to_u64(&from->source_timestamp);
    to->received_timestamp = dds_time_to_u64(&from->reception_timestamp);
  #else
    to->source_timestamp = 0;
    to->received_timestamp = 0;
  #endif /* !RTI_WIN32 */
```

Reply:
The question is, if we will change the clock to the `system_clock` how to deal with microseconds accuracy on Windows?

Reply:
> 1. The `system_clock` can jump backward after adjustment or clock synch. This could lead to the wrong order during replay.

Yes, that's true, but it is the only thing that makes sense in a distributed system.

> The question is, if we will change the clock to the `system_clock` how to deal with microseconds accuracy on Windows?

I don't necessarily think that is going to be a problem in Windows.  My recollection from looking into this a long time ago is that `std::chrono:system_clock` used the highest-accuracy clock it can find on the platform.  On Windows, this will likely be `QueryPerformanceFrequency`.  Indeed, in that blog post you link to, it shows that the `system_clock` on Windows has accuracy to 0.1 microseconds, i.e. 100 nanoseconds.  This is not as good as either `steady_clock`, or `high_resolution_clock`, but is definitely sufficient for our needs.

Reply:
Ok. Changing to the `system_clock` then.